crates/project_panel/src/project_panel.rs

```diff
@@ -279,9 +295,16 @@ impl ProjectPanel {
             .detach();
 
             let mut project_panel_settings = *ProjectPanelSettings::get_global(cx);
-            cx.observe_global::<SettingsStore>(move |_, cx| {
+            cx.observe_global::<SettingsStore>(move |this, cx| {
                 let new_settings = *ProjectPanelSettings::get_global(cx);
                 if project_panel_settings != new_settings {
+                    if new_settings.show_diagnostics != project_panel_settings.show_diagnostics {
+                        if new_settings.show_diagnostics != ShowDiagnostics::Off {
+                            this.update_diagnostics(cx);
+                        } else {
+                            this.diagnostics = Default::default();
+                        }
```

Comment:
Would be nice to restructure the code so you could always call `this.update_diagnostics(cx)` in case the value changed. Since `update_diagnostics` already resets the diagnostics, the rest of it would only need to be restructured so it only adds diagnostics based on the settings (which it already kinda does by using `show_warnings`)